crates/shared/src/event_handling.rs

```diff
@@ -161,24 +263,162 @@ where
         if !have_deleted_old_events {
             self.store.replace_events(Vec::new(), range.clone()).await?;
         }
-        self.last_handled_block = Some(range.end().to_u64());
+
+        let blocks = self
+            .block_retriever
+            .blocks(RangeInclusive::new(
+                range.end().saturating_sub(MAX_REORG_BLOCK_COUNT),
+                *range.end(),
+            ))
+            .await?;
+
+        self.update_last_handled_blocks(&blocks);
+        Ok(())
+    }
+
+    async fn update_events_from_latest_blocks(
+        &mut self,
+        blocks: &[BlockNumberHash],
+        is_reorg: bool,
+    ) -> Result<()> {
+        let (blocks, events) = self.past_events_by_block_hashes(blocks).await;
+        track_block_range(&format!("range_{}", blocks.len()));
+        tracing::debug!(
+            "final blocks for updating: {:?} - {:?}",
+            blocks.first(),
+            blocks.last()
+        );
+        if blocks.is_empty() {
+            return Ok(());
+        }
+        let range = RangeInclusive::new(blocks.first().unwrap().0, blocks.last().unwrap().0);
+        if is_reorg {
+            self.store.replace_events(events, range.clone()).await?;
+        } else {
+            self.store.append_events(events).await?;
+        }
+
+        self.update_last_handled_blocks(&blocks);
         Ok(())
     }
 
-    async fn past_events(
+    async fn past_events_by_block_hashes(
         &self,
-        block_range: &RangeInclusive<BlockNumber>,
+        blocks: &[BlockNumberHash],
+    ) -> (Vec<BlockNumberHash>, Vec<EthcontractEvent<C::Event>>) {
+        let (mut blocks_filtered, mut events) = (vec![], vec![]);
+        for (i, result) in future::join_all(
```

Comment:
I feel there should be a limit on the number of parallel calls. There are stream adapters in the futures crate that can do this.